parity-clib/parity.hpp

```diff
@@ -0,0 +1,228 @@
+    this->config = other.config;
+    other.config = nullptr;
+  }
+  ParityConfig &operator=(ParityConfig &&other) noexcept {
+    if (this != &other) {
+      this->config = other.config;
+      other.config = nullptr;
+    }
+    return *this;
+  }
+  ~ParityConfig() {
+    if (this->config)
+      assert(false && "ParityConfig objects must be moved into a "
+                      "ParityParams, not destroyed");
+  }
+};
+
+#define PARITY_DUMMY_CALLBACKS
+using parity_rpc_callback =
+    std::function<void(std::string_view const response)>;
+
+extern "C" void parity_cplusplus_callback(void *ud, const char *response,
+                                          size_t len) noexcept {
+#ifdef PARITY_DUMMY_CALLBACKS
+  (void)response, (void)len, (void)ud;
+#else
+  parity_rpc_callback *cb_ptr = static_cast<parity_rpc_callback *>(ud);
+  auto ptr = std::unique_ptr<parity_rpc_callback>(cb_ptr);
+  (*cb_ptr)(std::string_view(response, len));
+#endif
+};
+
+static inline void parity_subscription_callback(void *ud, const char *response,
+                                                size_t len) noexcept {
+#ifdef PARITY_DUMMY_CALLBACKS
+  (void)response, (void)len, (void)ud;
+#else
+  parity_rpc_callback *cb_ptr = static_cast<parity_rpc_callback *>(ud);
+  (*cb_ptr)(std::string_view(response, len));
+#endif
+};
+
+extern "C" void parity_destructor_callback(void *ud) noexcept {
+#ifdef PARITY_DUMMY_CALLBACKS
+  (void)ud;
+#else
+  if (ud)
+    delete static_cast<parity_rpc_callback *>(ud);
+#endif
+};
+
+typedef std::unique_ptr<::parity_subscription,
+                        decltype(parity_unsubscribe_ws) *>
+    parity_subscription;
+class ParityEthereum final {
+  struct ::parity_ethereum *parity_ethereum_instance;
+  parity_rpc_callback *callback;
+
+public:
+  ParityEthereum(const ParityEthereum &other) = delete;
+  ParityEthereum &operator=(const ParityEthereum &other) = delete;
+  ParityEthereum(ParityConfig config, ParityLogger logger,
+                 parity_rpc_callback new_chain_spec_callback)
+      : parity_ethereum_instance(nullptr),
+        callback(new parity_rpc_callback(new_chain_spec_callback)) {
+    struct ::ParityParams params = {
+        config.config,
+        parity_subscription_callback,
+        callback,
+        parity_destructor_callback,
+        logger.logger,
+    };
+    logger.logger = nullptr;
+    config.config = nullptr;
+    if (parity_start(&params, &this->parity_ethereum_instance))
+      throw std::runtime_error("Failed to start Parity Ethereum");
+  }
+  ParityEthereum(ParityEthereum &&other) noexcept
+      : parity_ethereum_instance(nullptr), callback(nullptr) {
+    this->parity_ethereum_instance = other.parity_ethereum_instance;
+    other.parity_ethereum_instance = nullptr;
+  }
+  ParityEthereum &operator=(ParityEthereum &&other) noexcept {
+    if (this != &other) {
+      this->parity_ethereum_instance = other.parity_ethereum_instance;
```

Comment:
I think you should call `parity_destroy(this->parity_ethereum_instance)` beforehand here. Same in the method above.

Reply:
Done